### DISCUSSION TITLE
What is gossipsub's equivalent of add_now_to_partial_view()?

### DISCUSSION THREAD

**Discussion by laptou** [libp2p/rust-libp2p#2193] [Q&A]
I'm trying to adapt the chat example [here](https://github.com/libp2p/rust-libp2p/blob/master/examples/chat.rs) (the one that uses Floodsub with mDNS) to use Gossipsub instead, but I can't figure it out. Here's what I tried:

```rust

#[derive(NetworkBehaviour)]
#[behaviour(out_event = "CombinedNetworkBehaviourEvent")]

struct CombinedNetworkBehaviour {
    mdns: Mdns,
    gossipsub: Gossipsub,
}

impl NetworkBehaviourEventProcess<GossipsubEvent> for CombinedNetworkBehaviour {
    fn inject_event(&mut self, event: GossipsubEvent) {
        debug!("received gossipsub event: {:?}", event);
    }
}

impl NetworkBehaviourEventProcess<MdnsEvent> for CombinedNetworkBehaviour {
    fn inject_event(&mut self, event: MdnsEvent) {
        debug!("received mdns event: {:?}", event);

        match event {
            MdnsEvent::Discovered(list) => {
                for (peer, addr) in list {
                    debug!("mdns: discovered peer {:?} at {:?}", peer, addr);
                    self.gossipsub.inject_connected(&peer)
                }
            }
            MdnsEvent::Expired(list) => {
                for (peer, addr) in list {
                    debug!("mdns: expired peer {:?} at {:?}", peer, addr);
                    self.gossipsub.inject_disconnected(&peer)
                }
            }
        }
    }
}

impl CombinedNetworkBehaviour {
    pub fn publish<H: libp2p::gossipsub::Hasher>(
        &mut self,
        topic: Topic<H>,
        data: impl Into<Vec<u8>>,
    ) -> Result<MessageId, libp2p::gossipsub::error::PublishError> {
        self.gossipsub.publish(topic, data)
    }
}

enum CombinedNetworkBehaviourEvent {
    Mdns(MdnsEvent),
    Gossip(GossipsubEvent),
}

impl From<MdnsEvent> for CombinedNetworkBehaviourEvent {
    fn from(e: MdnsEvent) -> Self {
        CombinedNetworkBehaviourEvent::Mdns(e)
    }
}

impl From<GossipsubEvent> for CombinedNetworkBehaviourEvent {
    fn from(e: GossipsubEvent) -> Self {
        CombinedNetworkBehaviourEvent::Gossip(e)
    }
}
```

But this doesn't work when I try to publish a message: I get the error that there are insufficient peers. 
I also tried using `add_explicit_peer()`, and it did work, but I'm not sure if it's actually doing what I want it to do.

**Comment by mxinden** [comment]
:wave:  Hi @laptou,
 
> ```rust
> impl NetworkBehaviourEventProcess<MdnsEvent> for CombinedNetworkBehaviour {
>     fn inject_event(&mut self, event: MdnsEvent) {
>         debug!("received mdns event: {:?}", event);
> 
>         match event {
>             MdnsEvent::Discovered(list) => {
>                 for (peer, addr) in list {
>                     debug!("mdns: discovered peer {:?} at {:?}", peer, addr);
>                     self.gossipsub.inject_connected(&peer)
> ```

`inject_connected` is a method on the `NetworkBehaviour` trait. As a rule of thumb, you will likely never need to call a method on the `NetworkBehaviour` trait as a user of rust-libp2p.

Instead, as you suggested `add_explicit_peer` could be an option, though that has additional implications, like staying connected to that peer at all times.

What I would do when building a basic example is connect to nodes discovered through mdns. The gossipsub subsystem will be notified of the new connection and then keep track of the peer. Something along the lines of `self.dial_addr(addr)`.


> ```rust
>             MdnsEvent::Expired(list) => {
>                 for (peer, addr) in list {
>                     debug!("mdns: expired peer {:?} at {:?}", peer, addr);
>                     self.gossipsub.inject_disconnected(&peer)
> ```

Same here, calling the _internal_ `inject_disconnected` method is not the way to go. If I understand your goal correctly, ignoring the address expiration can be ignored. 

> ```rust
> impl CombinedNetworkBehaviour {
>     pub fn publish<H: libp2p::gossipsub::Hasher>(
>         &mut self,
>         topic: Topic<H>,
>         data: impl Into<Vec<u8>>,
>     ) -> Result<MessageId, libp2p::gossipsub::error::PublishError> {
>         self.gossipsub.publish(topic, data)
> ```

Note: You could as well just call `swarm.behaviour_mut().gossip.publish(...)`.

---

Let us know if the above is of any help. Curious what you end up building.
